frontend/scripts/metrica.js

```diff
@@ -0,0 +1,231 @@
+
+        device.ios = () => {
+            return device.iphone() || device.ipod() || device.ipad();
+        };
+
+        device.iphone = () => {
+            return !device.windows() && find('iphone');
+        };
+
+        device.ipod = () => {
+            return find('ipod');
+        };
+
+        device.ipad = () => {
+            return find('ipad');
+        };
+
+        device.android = () => {
+            return !device.windows() && find('android');
+        };
+
+        device.androidPhone = () => {
+            return device.android() && find('mobile');
+        };
+
+        device.androidTablet = () => {
+            return device.android() && !find('mobile');
+        };
+
+        device.blackberry = () => {
+            return find('blackberry') || find('bb10') || find('rim');
+        };
+
+        device.blackberryPhone = () => {
+            return device.blackberry() && !find('tablet');
+        };
+
+        device.blackberryTablet = () => {
+            return device.blackberry() && find('tablet');
+        };
+
+        device.windows = () => {
+            return find('windows');
+        };
+
+        device.windowsPhone = () => {
+            return device.windows() && find('phone');
+        };
+
+        device.windowsTablet = () => {
+            return device.windows() && (find('touch') && !device.windowsPhone());
+        };
+
+        device.fxos = () => {
+            return (find('(mobile;') || find('(tablet;')) && find('; rv:');
+        };
+
+        device.fxosPhone = () => {
+            return device.fxos() && find('mobile');
+        };
+
+        device.fxosTablet = () => {
+            return device.fxos() && find('tablet');
+        };
+
+        device.meego = () => {
+            return find('meego');
+        };
+
+        device.cordova = () => {
+            return window.cordova && location.protocol === 'file:';
+        };
+
+        device.nodeWebkit = () => {
+            return typeof window.process === 'object';
+        };
+
+        device.mobile = () => {
+            return device.androidPhone() ||
+                device.iphone() ||
+                device.ipod() ||
+                device.windowsPhone() ||
+                device.blackberryPhone() ||
+                device.fxosPhone() ||
+                device.meego();
+        };
+
+        device.tablet = () => {
+            return device.ipad() ||
+                device.androidTablet() ||
+                device.blackberryTablet() ||
+                device.windowsTablet() ||
+                device.fxosTablet();
+        };
+
+        device.desktop = () => {
+            return !device.tablet() && !device.mobile();
+        };
+
+        find = (needle) => {return userAgent.indexOf(needle) !== -1};
+
+        return device;
+    };
+
+    const getUserAgent = () => {
+        return window.navigator.userAgent;
+    };
+
+    const setToken = (token) => {
+        localStorage.setItem('visitor_token', token);
```

Comment:
localStorage not always accessible, see as it is done inside the platform metrica. There should be some storageService

Reply:
and wrap such things with try catch, in order not to have undesirable exceptions